### PR TITLE
feat(bazel): Use auto-replace when possible

### LINKED ISSUE #9667: Refactor Bazel to use autoReplace + artifacts
I think we should refactor our Bazel handling to use getArtifacts instead of trying to update them all in one go using updateDependency().

Whenever URLs, digests or checksums are dependent on the version - and not independently updatable - it should usually mean we use artifacts.

### PR DESCRIPTION
**Pull request by zharinov**
## Changes

Use auto-replace when possible and use artifacts update `http` rules where needed.

## Context

- Closes: #9667

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/manager/bazel/rules/go.ts [hunk 1]
```diff
@@ -26,10 +26,12 @@ export const GoTarget = z
...
-      }
-
-      if (commit) {
-        dep.currentValue = 'v0.0.0';
+        if (commit) {
+          dep.currentDigest = commit;
+          dep.currentDigestShort = commit.substring(0, 7);
+        }
+      } else if (commit) {
+        dep.currentValue = tag;
```
Comment (viceice) [inline lib/modules/manager/bazel/rules/go.ts:34]:
```suggestion
```
tag is undefined here, so no need to return

File: lib/modules/manager/bazel/artifacts.ts [hunk 2]
```diff
@@ -171,16 +140,19 @@ export async function updateDependency({
...
-      let result = fileContent;
-      result = patchCodeAtFragments(result, urlFragments, updateValues);
-      result = updateCode(result, [idx, 'strip_prefix'], updateValues);
-      result = updateCode(result, [idx, 'sha256'], hash);
-      return result;
+      contents = patchCodeAtFragments(contents, urlFragments, updateValues);
+      contents = updateCode(contents, [idx, 'strip_prefix'], updateValues);
+      contents = updateCode(contents, [idx, 'sha256'], hash);
     }
-  } catch (err) /* istanbul ignore next */ {
```
Comment (viceice) [inline lib/modules/manager/bazel/artifacts.ts:180]: shouldn't we still catch errors?

Comment (zharinov) [reply lib/modules/manager/bazel/artifacts.ts:180]: Good catch

Comment (zharinov) [reply lib/modules/manager/bazel/artifacts.ts:180]: If I get it right, when we throw error from artifact update function, it means retry for the next run. When we return `artifactError` field, it means PR warning and retry can be forced manually by checking the checkbox.

The only thing that can throw here I believe is the HTTP request, but it's already catched. So I think this `catch` clause is redundant.

Comment (zharinov) [reply lib/modules/manager/bazel/artifacts.ts:180]: Maybe let's just see which errors we can encounter here and adjust to more desired behavior eventually

File: lib/modules/manager/bazel/rules/go.ts [hunk 3]
```diff
@@ -26,10 +26,11 @@ export const GoTarget = z
 
       if (tag) {
         dep.currentValue = tag;
-      }
-
-      if (commit) {
-        dep.currentValue = 'v0.0.0';
+        if (commit) {
+          dep.currentDigest = commit;
+          dep.currentDigestShort = commit.substring(0, 7);
```
Comment (viceice) [inline lib/modules/manager/bazel/rules/go.ts:31]: isn't this done inside worker when preparing the updates?

File: lib/modules/manager/bazel/artifacts.spec.ts [hunk 4]
```diff
@@ -0,0 +1,554 @@
+import crypto from 'crypto';
+import { codeBlock } from 'common-tags';
+import * as httpMock from '../../../../test/http-mock';
+import { partial } from '../../../../test/util';
+import type { UpdateArtifact } from '../types';
+import { updateArtifacts } from './artifacts';
```
Comment (viceice) [inline lib/modules/manager/bazel/artifacts.spec.ts:6]:
```suggestion
import { updateArtifacts } from '.';
```
better code coverage for branches

**Review by rarkins — DISMISSED** [review]
Let's land this in v35?

**Comment by zharinov** [comment]
🤷

File: lib/modules/manager/bazel/rules/go.ts [hunk 5]
```diff
@@ -29,10 +29,10 @@ export const GoTarget = z
       }
 
       if (commit) {
-        dep.currentValue = 'v0.0.0';
         dep.currentDigest = commit;
-        dep.currentDigestShort = commit.substring(0, 7);
-        dep.digestOneAndOnly = true;
+        if (!tag) {
+          dep.digestOneAndOnly = true;
```
Comment (viceice) [inline lib/modules/manager/bazel/rules/go.ts:34]: 🤔 where is that property used? never seen it before

Comment (zharinov) [reply lib/modules/manager/bazel/rules/go.ts:34]: Only with `go`, and actually I'm not sure what it's for exactly

Comment (rarkins) [reply lib/modules/manager/bazel/rules/go.ts:34]: It was a hack when I added Go. I forget exactly why or if it's still necessary

Comment (zharinov) [reply lib/modules/manager/bazel/rules/go.ts:34]: I saw no difference with or without it

Comment (rarkins) [reply lib/modules/manager/bazel/rules/go.ts:34]: Let's keep it for now

**Review by viceice — APPROVED** [review]

**Pull request merged**